### PR TITLE
Framework for Programmable Tab Completion

### PR DESCRIPTION
**Pull request by adqm**
This builds on #1027 and drastically improves start-up times (except on the first run) by caching information about bash completion functions.

### CODE REVIEW

File: docs/tutorial_completers.rst [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+xonsh actually uses, in the ``xonsh.completers`` module.
+
+.. code-block:: python
+
+    def dummy_completer(prefix, line, begidx, endidx, ctx):
+        '''
+        Completes everything with options "lou" and "carcolh",
+        regardless of the value of prefix.
+        '''
+        return {"lou", "carcolh"}
```
Comment (gforsyth) [inline docs/tutorial_completers.rst]: :grin:

File: xonsh/completers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from xonsh.completers.python import (complete_python, complete_import,
+                                     complete_python_mode)
+from xonsh.completers.commands import complete_skipper
+
+completers = OrderedDict()
+completers['python_mode'] = complete_python_mode
+completers['base'] = complete_base
+completers['skip'] = complete_skipper
+completers['cd'] = complete_cd
+completers['rmdir'] = complete_cd
```
Comment (gforsyth) [inline xonsh/completers/__init__.py]: this should be `complete_rmdir`, no?

Comment (adqm) [inline xonsh/completers/__init__.py]: Oops...

Yes, indeed.

File: docs/tutorial_completers.rst [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+================================
+
+As with many other shells, xonsh ships with the ability to complete
+partially-specified arguments upon hitting the "tab" key.
+
+In Python-mode, pressing the "tab" key will complete based on the variable
+names in the current builtins, globals, and locals, as well as xonsh language
+keywords & operators, files & directories, and environment variable names. In
+subprocess-mode, xonsh additionally completes based on the names of any
+executable files on your $PATH, alias keys, and full BASH completion for the
```
Comment (scopatz) [inline docs/tutorial_completers.rst]: `Bash` is the appropriate spelling, like `Python`

File: docs/tutorial_completers.rst [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+* ``"<KEY"``, where ``KEY`` is a pre-existing name, indicates that this should be added before the completer named ``KEY``
+
+.. note:: It is also possible to manipulate ``__xonsh_completers__`` directly,
+          but this is the preferred method.
+
+Removing a Completer
+====================
+
+To remove a completer from the list of active completers, run
+``remove-completer NAME``, where ``NAME`` is the unique identifier associated
```
Comment (scopatz) [inline docs/tutorial_completers.rst]: I guess I would prefer an interface that was `completer add` and `completer rm` or similar than two different commands.  I'd be happy to modify this later, though, and should not be blocking here, if you are amicable.

File: xonsh/ast.py [hunk 5]
```diff
@@ -92,7 +92,7 @@ class CtxAwareTransformer(NodeTransformer):
     a subprocess.
     """
 
-    def __init__(self, parser):
+    def __init__(self, parser, wrap_subprocs=True):
```
Comment (scopatz) [inline xonsh/ast.py]: I don't know if this makes sense.  This flag basically disables the class from performing any work. Rather it seems that whatever call this (the execer) should simply made to skip this, if desired. Even if this is desired here, I believe that this would be better as part of the kwargs passed to the methods, and not part of the instance state.

File: xonsh/completer.py [hunk 6]
```diff
@@ -163,448 +36,15 @@ def complete(self, prefix, line, begidx, endidx, ctx=None):
...
-        except:
-            self._options = {}
-
-    def _save_cached_options(self):
-        """Save completions to file."""
-        with open(OPTIONS_PATH, 'wb') as f:
-            pickle.dump(self._options, f)
+                res = out
+                lprefix = len(prefix)
+            if res is not None and res != empty_set:
```
Comment (scopatz) [inline xonsh/completer.py]: Isn't it better to test if `len(res) > 0`, rather than creating an extra set instance?

File: xonsh/completers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+completers['python_mode'] = complete_python_mode
+completers['base'] = complete_base
+completers['skip'] = complete_skipper
+completers['cd'] = complete_cd
+completers['rmdir'] = complete_rmdir
+completers['bash'] = complete_from_bash
+completers['man'] = complete_from_man
+completers['import'] = complete_import
+completers['python'] = complete_python
+completers['path'] = complete_path
```
Comment (scopatz) [inline xonsh/completers/__init__.py]: I believe that you can pass in `OrderedDict([('python_mode', complete_python_mode), ('base', complete_base), ...])` instead.

File: xonsh/completers/__init__.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+builtins.__xonsh_completers__ = completers
```
Comment (scopatz) [inline xonsh/completers/__init__.py]: Also this should happen in `xonsh.built_ins` and also be unloaded.

File: xonsh/completers/_aliases.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    if loc == 'start':
+        new[name] = func
+        for (k, v) in builtins.__xonsh_completers__.items():
+            new[k] = v
+    elif loc == 'end':
+        for (k, v) in builtins.__xonsh_completers__.items():
+            new[k] = v
+        new[name] = func
+    else:
+        dir, rel = loc[0], loc[1:]
```
Comment (scopatz) [inline xonsh/completers/_aliases.py]: Maybe use a name other than the builtin `dir` :)

File: xonsh/completers/_aliases.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            if rel == k and dir == '<':
+                new[name] = func
+                found = True
+            new[k] = v
+            if rel == k and dir == '>':
+                new[name] = func
+                found = True
+        if not found:
+            new[name] = func
+    builtins.__xonsh_completers__ = new
```
Comment (scopatz) [inline xonsh/completers/_aliases.py]: Maybe a clear() and an update() would be better than complete replacement.  That way, folks that have a reference to the original object would not loose it.  Namely,

```python
builtins.__xonsh_completers__.clear()
builtins.__xonsh_completers__.update(new)
```

You could also change new to just be a list of tuples.

File: xonsh/completers/commands.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+from xonsh.tools import executables_in
+from xonsh.platform import ON_WINDOWS
+from xonsh.completers.tools import get_filter_function
+
+SKIP_TOKENS = {'sudo', 'time', 'timeit', 'which', 'showcmd', 'man'}
+
+
+def complete_command(cmd, line, start, end, ctx):
+    space = ' '
```
Comment (scopatz) [inline xonsh/completers/commands.py]: Should have a docstring.

File: xonsh/completers/dirs.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import os
+
+from xonsh.completers.man import complete_from_man
+from xonsh.completers.path import complete_dir
+
+PREVENT_OTHERS = ['path']
```
Comment (scopatz) [inline xonsh/completers/dirs.py]: What is this used for?

File: xonsh/completers/man.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
+import os
+import re
+import pickle
+import subprocess
+
+from xonsh.completers.tools import get_filter_function
+
+OPTIONS_PATH = os.path.join(os.path.expanduser('~'), ".xonsh_man_completions")
```
Comment (scopatz) [inline xonsh/completers/man.py]: I think we should change this location to the `$XONSH_DATA_DIR`

File: xonsh/completers/python.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+XONSH_TOKENS = {
+    'and ', 'as ', 'assert ', 'break', 'class ', 'continue', 'def ', 'del ',
+    'elif ', 'else', 'except ', 'finally:', 'for ', 'from ', 'global ',
+    'import ', 'if ', 'in ', 'is ', 'lambda ', 'nonlocal ', 'not ', 'or ',
+    'pass', 'raise ', 'return ', 'try:', 'while ', 'with ', 'yield ', '+', '-',
+    '/', '//', '%', '**', '|', '&', '~', '^', '>>', '<<', '<', '<=', '>', '>=',
+    '==', '!=', '->', '=', '+=', '-=', '*=', '/=', '%=', '**=', '>>=', '<<=',
+    '&=', '^=', '|=', '//=', ',', ';', ':', '?', '??', '$(', '${', '$[', '..',
+    '...', '![', '!(',
```
Comment (scopatz) [inline xonsh/completers/python.py]: `@` and `@(` should probably be added here.

File: xonsh/completers/tools.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+import builtins
+
+
+def get_filter_function():
```
Comment (scopatz) [inline xonsh/completers/tools.py]: This name doesn't make a lot of sense to me.  Also, wouldn't it be faster to have the two other functions defined globally and then simply return one or the other.  These get redefined each time this function is called.

File: xonsh/completers/tools.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    else:
+        def filt(s, x): return s.lower().startswith(x.lower())
+    return filt
+
+
+def is_iterable(x):
+    try:
+        _ = iter(x)
+        return True
+    except:
```
Comment (scopatz) [inline xonsh/completers/tools.py]: This should probably yield a type error.  Also,  I think the prefered version of this function is simply `isinstance(x, collections.abs.Iterable)`

File: xonsh/completers/tools.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+        return False
+
+
+def justify(s, max_length, left_pad=0):
```
Comment (scopatz) [inline xonsh/completers/tools.py]: Please provide a docstring.  Also, this looks suspiciously similar to tools that are already available in `textwrap` in the standard library.

Comment (adqm) [inline xonsh/completers/tools.py]: Oh, cool.  I had no idea that module existed (I've used this piece of code before...).  You're right I can probably just use that.

File: xonsh/completers/tools.py (hunk 16 above)
Comment (adqm) [inline xonsh/completers/tools.py]: Again, no idea that one existed.  I'll replace throughout.

File: xonsh/completers/tools.py (hunk 15 above)
Comment (adqm) [inline xonsh/completers/tools.py]: It's the function that is used to `filter` the list, was my reason for calling it that.  Agree about defining them globally.  Will make that change.

File: xonsh/completers/python.py (hunk 14 above)
Comment (adqm) [inline xonsh/completers/python.py]: Agreed.  Will do.

File: xonsh/execer.py [hunk 18]
```diff
@@ -68,7 +68,8 @@ def parse(self, input, ctx, mode='exec'):
         # tokens for all of the Python rules. The lazy way implemented here
         # is to parse a line a second time with a $() wrapper if it fails
         # the first time. This is a context-free phase.
-        tree, input = self._parse_ctx_free(input, mode=mode)
+        tree, input = self._parse_ctx_free(input, mode=mode,
+                                           wrap_subprocs=wrap_subprocs)
```
Comment (scopatz) [inline xonsh/execer.py]: So here is where I think that if `wrap_subprocs` is False, it should just 

```
return self.parser.parse(input, filename=self.filename, mode=mode, debug_level=(self.debug_level > 1))
```

Rather than adding arguments to this and the following function.

File: xonsh/completers/man.py (hunk 13 above)
Comment (adqm) [inline xonsh/completers/man.py]: Yep, agreed.

File: xonsh/completers/dirs.py (hunk 12 above)
Comment (adqm) [inline xonsh/completers/dirs.py]: Nothing.  A holdover from an older version.  I'll kill it.

File: xonsh/completers/commands.py (hunk 11 above)
Comment (adqm) [inline xonsh/completers/commands.py]: Yep, I'll add one.

File: xonsh/execer.py [hunk 19]
```diff
@@ -122,12 +127,13 @@ def exec(self, input, mode='exec', glbs=None, locs=None, stacklevel=2):
...
                                 mode=mode,
-                                stacklevel=stacklevel)
+                                stacklevel=stacklevel,
+                                wrap_subprocs=wrap_subprocs)
         if code is None:
             return None  # handles comment only input
         return exec(code, glbs, locs)
 
-    def _parse_ctx_free(self, input, mode='exec'):
+    def _parse_ctx_free(self, input, mode='exec', wrap_subprocs=True):
```
Comment (scopatz) [inline xonsh/execer.py]: Yeah, I don' think that `wrap_subprocs` is needed here.

File: xonsh/execer.py [hunk 20]
```diff
@@ -144,6 +150,8 @@ def _parse_ctx_free(self, input, mode='exec'):
                 else:
                     raise original_error
             except SyntaxError as e:
+                if not wrap_subprocs:
+                    raise e
```
Comment (scopatz) [inline xonsh/execer.py]: ... which will allow these two line to be removed.

File: xonsh/completers/_aliases.py (hunk 10 above)
Comment (adqm) [inline xonsh/completers/_aliases.py]: Oh, yes.  Good call.

File: xonsh/completers/_aliases.py (hunk 9 above)
Comment (adqm) [inline xonsh/completers/_aliases.py]: Of course, can do.

File: xonsh/completers/__init__.py (hunk 8 above)
Comment (adqm) [inline xonsh/completers/__init__.py]: That was what I had done originally, but, IIRC, I had a problem with circular imports.  I think I still had the code to unload it on exit, but initializing it here was easier.  I can try to see if there's a better way...

File: xonsh/main.py [hunk 21]
```diff
@@ -190,7 +191,8 @@ def premain(argv=None):
                     'completer': False,
                     'login': False,
                     'scriptcache': args.scriptcache,
-                    'cacheall': args.cacheall}
+                    'cacheall': args.cacheall,
+                    'ctx': builtins.__xonsh_ctx__}
```
Comment (scopatz) [inline xonsh/main.py]: I am not sure why this change is needed...

File: xonsh/completers/__init__.py (hunk 7 above)
Comment (adqm) [inline xonsh/completers/__init__.py]: Yes, indeed.

File: xonsh/completer.py (hunk 6 above)
Comment (adqm) [inline xonsh/completer.py]: Yes, that would make some sense.

File: docs/tutorial_completers.rst (hunk 4 above)
Comment (adqm) [inline docs/tutorial_completers.rst]: You're right; that's probably a nicer interface.  I can hack that in here.

**Comment by scopatz** [comment]
This is a huge feat @adqm! Thanks a million for putting this in!  Also, thanks in adavance for all of my questions.

File: docs/tutorial_completers.rst (hunk 4 above)
Comment (scopatz) [inline docs/tutorial_completers.rst]: Thanks!

File: xonsh/completers/__init__.py (hunk 8 above)
Comment (scopatz) [inline xonsh/completers/__init__.py]: Yeah, I think we need to streamline and simplify the startup process. Things are kind of a tangled mess right now. (My bad)

File: xonsh/completers/dirs.py (hunk 12 above)
Comment (scopatz) [inline xonsh/completers/dirs.py]: Ahh great!  I thought I was missing something

File: xonsh/execer.py (hunk 18 above)
Comment (adqm) [inline xonsh/execer.py]: Oooooooh.  Okay.  I was confused by your comments about this initially, but this cleared it up.  I can make that change.

File: xonsh/completers/tools.py (hunk 16 above)
Comment (scopatz) [inline xonsh/completers/tools.py]: Thanks!  And, I meant `collections.abc.Iterable`.  Sorry about the typo

File: xonsh/main.py (hunk 21 above)
Comment (adqm) [inline xonsh/main.py]: This was needed to make sure that, if `register-completer` is invoked inside the xonshrc, that it actually has access to the functions that are defined there (since `register-completer` looks in `builtins.__xonsh_ctx__` for them).

File: xonsh/completers/tools.py (hunk 17 above)
Comment (scopatz) [inline xonsh/completers/tools.py]: Yeah, it is worth looking into :).  Anytime we can push code burden into the standard library, we probably should!

**Comment by adqm** [comment]
Thanks for the detailed review!  I agree with all your comments.  I'll go through this in detail, hopefully tonight or tomorrow.

File: xonsh/completers/tools.py (hunk 17 above)
Comment (adqm) [inline xonsh/completers/tools.py]: I had a hard time making `textwrap` do what I wanted it to, so I would prefer to stick with this for now if that works for you.

**Comment by adqm** [comment]
Alright @scopatz, I think I hit all your comments, and fixed another bug in the process.  Should be ready for another look!

File: xonsh/main.py (hunk 21 above)
Comment (scopatz) [inline xonsh/main.py]: OK, thanks for the explanation.

**Comment by scopatz** [comment]
Thanks @adqm! The code looks good to me. Playing around with this now.  Would you mind adding API autodoc hooks? Or we can make another issue for that....

**Comment by scopatz** [comment]
Also, thanks to @gforsyth who has snuck in some commits :)

**Comment by scopatz** [comment]
OK I am going to make an executive decision and merge this. It is freaking amazing! I'll add the API docs directly.

**Comment by scopatz** [comment]
Thanks to everyone who helped with this one!

**Comment by adqm** [comment]
Woo-hoo!  We finally made it!  Thanks to all for help with testing :)
